### ISSUE TITLE
Transient error with mc/edit-beginnings-of-lines

### ISSUE THREAD

**Issue by hmelman** [magit/transient#259]
I've created a transient with many multiple-cursor commands.  He's a small form of it to recreate the problem.

```elisp
(require 'multiple-cursors)

(transient-define-prefix hrm-mc-transient ()
  "Multiple-cursor commands"
  :transient-suffix 'transient--do-exit
  ["Mark"
   ["Lines"
    ("ll" "Column" mc/edit-lines)
    ("la" "Beginning" mc/edit-beginnings-of-lines)
    ("le" "End" mc/edit-ends-of-lines)
    ]
   ]
  )

(global-set-key (kbd "C-c C-m") 'hrm-mc-transient)
```

I find setting a region and using `ll` works fine, but `la` results in this backtrace (and `le` is similar):

```
Debugger entered--Lisp error: (wrong-type-argument (or eieio-object cl-structure-object oclosure) nil)
  apply(debug (error (wrong-type-argument (or eieio-object cl-structure-object oclosure) nil)))
  transient--exit-and-debug(error (wrong-type-argument (or eieio-object cl-structure-object oclosure) nil))
  signal(wrong-type-argument ((or eieio-object cl-structure-object oclosure) nil))
  eieio-oref(nil transient-non-suffix)
  transient--get-predicate-for(beginning-of-line)
  transient--do-pre-command()
  transient--pre-command()
  run-hooks(pre-command-hook)
  mc/execute-command(beginning-of-line)
  mc/execute-command-for-fake-cursor(beginning-of-line #<overlay in no buffer>)
  #f(compiled-function (cursor) #<bytecode 0x220d79323fa2437>)(#<overlay in no buffer>)
  mapc(#f(compiled-function (cursor) #<bytecode 0x220d79323fa2437>) (#<overlay in no buffer> #<overlay from 248 to 249 in fallfocus.md> #<overlay from 553 to 553 in fallfocus.md> #<overlay from 554 to 555 in fallfocus.md>))
  mc/execute-command-for-all-fake-cursors(beginning-of-line)
  mc/execute-command-for-all-cursors(beginning-of-line)
  #f(compiled-function () (interactive nil) #<bytecode -0xbbc8b4969ac642a>)()
  apply(#f(compiled-function () (interactive nil) #<bytecode -0xbbc8b4969ac642a>) nil)
  #f(compiled-function (fn &rest args) #<bytecode -0x103378276b2fcaef>)(#f(compiled-function () (interactive nil) #<bytecode -0xbbc8b4969ac642a>))
  apply(#f(compiled-function (fn &rest args) #<bytecode -0x103378276b2fcaef>) #f(compiled-function () (interactive nil) #<bytecode -0xbbc8b4969ac642a>) nil)
  (lambda (fn &rest args) (interactive #f(compiled-function (spec) #<bytecode 0xec321c284c529e7>)) (apply '#f(compiled-function (fn &rest args) #<bytecode -0x103378276b2fcaef>) fn args))(#f(compiled-function () (interactive nil) #<bytecode -0xbbc8b4969ac642a>))
  apply((lambda (fn &rest args) (interactive #f(compiled-function (spec) #<bytecode 0xec321c284c529e7>)) (apply '#f(compiled-function (fn &rest args) #<bytecode -0x103378276b2fcaef>) fn args)) #f(compiled-function () (interactive nil) #<bytecode -0xbbc8b4969ac642a>) nil)
  mc/edit-beginnings-of-lines()
  funcall-interactively(mc/edit-beginnings-of-lines)
  call-interactively(mc/edit-beginnings-of-lines nil nil)
  command-execute(mc/edit-beginnings-of-lines)
```

Magit 20230924.1834 [>= 3.3.0.50-git], Transient 0.4.3, Git 2.40.0, Emacs 29.1, darwin
I'm using the macport of emac 29 and Multiple cursors: 1.4.0 (from melpa multiple-cursors-20230728)

I'm guessing it has something to do with the fact that `mc/edit-beginnings-of-lines` calls the command `beginning-of-line` and this is happening with the transient still enabled and confusing things, but I'm really not sure.  Any hints to try to track this down would be appreciated.

**Comment by tarsius** [comment]
By default Transient does not allow calling commands that are not explicitly added as suffix commands. (Additionally some common commands, which are part of Transient itself, are always allowed.)

If a suffix commands uses `call-interactively` then that qualifies as calling a command.

You can allow all non-suffix commands to be called by adding `:transient-non-suffix` to the prefix definition.

**Comment by hmelman** [comment]
Thanks.  I don't completely follow but I tried this and it failed the same way:

```elisp
(transient-define-prefix hrm-mc-transient ()
  "Multiple-cursor commands"
  :transient-non-suffix t
  ["Mark"
   ["Lines"
    ("ll" "Column" mc/edit-lines)
    ("la" "Beginning" mc/edit-beginnings-of-lines)
    ("le" "End" mc/edit-ends-of-lines)
    ]
   ]
  )
```

**Comment by tarsius** [comment]
Try also adding `:transient-suffix t`. That might not be the behavior you want, but let's try it anyway.

**Comment by hmelman** [comment]
Yes, that does make all three commands work.  As you suspect, it does mean the transient stays active after the command runs and this is not desired.

**Comment by tarsius** [comment]
(from https://github.com/magnars/multiple-cursors.el#maintenance-warning)

> TLDR: I am still maintaining this package, but I am no longer crowdsourcing a list of issues.

I'm so gonna steal that line. :grin: 

@magnars, I am not trying to trick you into doing any work!

----

Since https://github.com/magnars/multiple-cursors.el/commit/8c8d52f248324b39f68eb8add96353e2b1bb07b2 `multiple-cursors` explicitly calls `{pre,post}-command-hook`. It probably shouldn't. Anyway, both `multiple-cursors` and `transient` do funky stuff with these hooks, and they conflict.

You can disable these hooks as shown below. In most cases that probably won't cause any issues, but I don't know for sure -- the relevant commit messages didn't mention why these hooks are being called (at least not the ones I looked at).

```lisp
(defun test-suffix ()
  (interactive)
  (setq this-command 'beginning-of-line) ;questionable
  (call-interactively 'beginning-of-line))

(defun test-mc ()
  (interactive)
  (mc/execute-command 'beginning-of-line))

(defun test-mc-clean-hook ()
  (interactive)
  ;; doesn't work for some reason
  ;; (let ((pre-command-hook (remq 'transient--pre-command pre-command-hook))
  ;;       (post-command-hook (remq 'transient--post-command post-command-hook)))
  (let ((pre-command-hook nil)
        (post-command-hook nil))
    (mc/execute-command 'beginning-of-line)))

(transient-define-prefix test-prefix ()
  :transient-non-suffix t
  [("p" "plain" test-suffix)
   ("m" "multi" test-mc)
   ("c" "multi clean hook" test-mc-clean-hook)])

;; (test-prefix)
```

**Comment by magnars** [comment]
In order to work with all kinds of home-made commands and modes, multiple-cursors needs to simulate the core Emacs command loop once for each cursor, which includes the pre and post hooks. So yeah, the entire multiple-cursors package lands pretty squarely in the "probably shouldn't" box, but for some reason it's been pretty stable the last 11 years. It's a wonder to me that it doesn't break like this more often.

For most of those years, I had a section in the README about how multiple-cursor commands needed to be bound to keys, and couldn't be used with M-x, but we got a fix for that a few years ago. In that defeatist spirit, maybe one possible solution in this case is to just not use transient with the multiple-cursors commands?

**Comment by tarsius** [comment]
It should be possible to do something about this, but I am also very busy right now.

@hmelman I'm afraid you'll have to wait some more.

**Comment by tarsius** [comment]
Try

```lisp
(defun mc@transient--pre-command (fn)
  (unless mc--executing-command-for-fake-cursor
    (funcall fn)))
(advice-add 'transient--pre-command :around #'mc@transient--pre-command)

(defun mc@transient--post-command (fn)
  (unless mc--executing-command-for-fake-cursor
    (funcall fn)))
(advice-add 'transient--post-command :around #'mc@transient--post-command)
```

This assumes that `mc/execute-command-for-all-fake-cursors` is used, not `mc/execute-command`.

**Comment by tarsius** [comment]
I've added a [FAQ](https://github.com/magit/transient/wiki/FAQ) entry about this
